crypto/armor.go

```diff
@@ -193,16 +212,33 @@ func UnarmorDecryptPrivKey(armorStr string, passphrase string) (privKey cryptoty
 	return privKey, header[headerType], err
 }
 
-func decryptPrivKey(saltBytes []byte, encBytes []byte, passphrase string) (privKey cryptotypes.PrivKey, err error) {
-	key, err := bcrypt.GenerateFromPassword(saltBytes, []byte(passphrase), BcryptSecurityParameter)
-	if err != nil {
-		return privKey, errorsmod.Wrap(err, "error generating bcrypt key from passphrase")
+func decryptPrivKey(saltBytes []byte, encBytes []byte, passphrase string, kdf string) (privKey cryptotypes.PrivKey, err error) {
+	// Key derivation
+	var key []byte
+	switch kdf {
+	case KDFArgon2:
+		key = argon2.IDKey([]byte(passphrase), saltBytes, argon2Time, argon2Memory, argon2Threads, argon2KeyLen)
+	default:
+		key, err = bcrypt.GenerateFromPassword(saltBytes, []byte(passphrase), BcryptSecurityParameter)
+		if err != nil {
+			return privKey, errorsmod.Wrap(err, "error generating bcrypt key from passphrase")
+		}
+		key = crypto.Sha256(key) // Get 32 bytes
 	}
 
-	key = crypto.Sha256(key) // Get 32 bytes
+	// Symetrhic decryption
+	aead, err := chacha20poly1305.New(key)
+	if len(encBytes) < aead.NonceSize() {
+		return privKey, errorsmod.Wrap(err, "error generating aead cypher from key")
+	}
+	nonce, privKeyBytesEncrypted := encBytes[:aead.NonceSize()], encBytes[aead.NonceSize():] // Split nonce and ciphertext.
+	privKeyBytes, err := aead.Open(nil, nonce, privKeyBytesEncrypted, nil)                   // Decrypt the message and check it wasn't tampered with.
+	// Fallback: tries decryption with legacy encrypthin algorithm "xsalsa20symmetric"
+	if err != nil {
+		privKeyBytes, err = xsalsa20symmetric.DecryptSymmetric(encBytes, key)
```

Comment:
If key was computed as `key = argon2.IDKey([]byte(passphrase), saltBytes, argon2Time, argon2Memory, argon2Threads, argon2KeyLen)` this operation doesn't make sense, you have the wrong key.

If decryption with the Argon2id derived key fails, then re-generate a key as in the `default` case, and try `privKeyBytes, err = xsalsa20symmetric.DecryptSymmetric(encBytes, key)`

You could put all the logic for Argon2id key derivation and `chacha20poly1305` decryption in the `Argon2id` case. And if decryption fails, then fall through to the default case (i.e. bcrypt key derivation and `xsalsa20symmetric` decryption)